File: refrence-service/src/main/java/com/kapitonau/ps/refrenceservice/repository/ReferenceItemRepository.java
package com.kapitonau.ps.refrenceservice.repository;

import com.kapitonau.ps.refrenceservice.model.ReferenceItemModel;
import org.springframework.data.jpa.repository.JpaRepository;
import org.springframework.data.jpa.repository.Query;

import java.util.List;
import java.util.Optional;

public interface ReferenceItemRepository extends JpaRepository<ReferenceItemModel, Long> {

    @Query("""
            select ri from ReferenceItemModel ri 
                        where ri.isActive and ri.referenceType = :referenceType 
                                    and (:itemCode is null or ri.itemCode = :itemCode)
            """)
    List<ReferenceItemModel> findAllByReferenceTypeAndItemCode(String referenceType, String itemCode);


    @Query("""
        select ri from ReferenceItemModel ri where ri.isActive and ri.referenceItemId = :referenceItemId
        """)
    Optional<ReferenceItemModel> findByReferenceItemId(Long referenceItemId);

}

File: reference-service/src/main/kotlin/com/kapitonau/referenceservice/repository/ReferenceItemRepository.kt
package com.kapitonau.referenceservice.repository

import com.kapitonau.referenceservice.model.ReferenceItemModel
import org.springframework.data.jpa.repository.JpaRepository
import org.springframework.data.jpa.repository.Query
import org.springframework.data.repository.query.Param

interface ReferenceItemRepository : JpaRepository<ReferenceItemModel, Long>, CustomReferenceItemRepository {

    @Query(
        """
    SELECT i FROM ReferenceItemModel i 
    WHERE (:referenceType IS NULL OR i.referenceType.referenceType = :referenceType)
    AND (:itemCode IS NULL OR i.itemCode = :itemCode)
"""
    )
    fun findAllByReferenceTypeAndItemCode(
        @Param("referenceType") referenceType: String?,
        @Param("itemCode") itemCode: String?
    ): List<ReferenceItemModel>

}